Fill new `protocol` field in a `Tunnel` record if the info is present in the tunnel's network's other config.

In XCP-ng before 8.2, our open-vxlan-port-for-sdn-controller made XAPI open the vxlan port automatically when
there was a SDN controller, independantly of the protocol (GRE or vxlan).

Starting with XCP-ng 8.2, the new open-vxlan-port-for-sdn-controller patch (based on the PR merged upstream
after stockholm's release [CH 8.2]) adds a new `Tunnel` record with a `protocol` field and only opens the port
if the protocol is vxlan.

In order to still support tunnels created before `Tunnel` was added, the present patch leverages the XAPI database
upgrade mechanism to create the `Tunnel` objects when a tunnel is found, reading the protocol information from
the network's `other_config` field where XO used to put it before the change.

This patch will remain needed as long as we want to support such smooth transition for tunnels created before
XCP-ng 8.2.

diff --git i/ocaml/xapi/xapi_db_upgrade.ml w/ocaml/xapi/xapi_db_upgrade.ml
index 05f20f083..3aaf3fead 100644
--- i/ocaml/xapi/xapi_db_upgrade.ml
+++ w/ocaml/xapi/xapi_db_upgrade.ml
@@ -106,6 +106,10 @@ let jura =
   , Datamodel_common.jura_release_schema_minor_vsn
   )

+let quebec =
+  ( Datamodel_common.quebec_release_schema_major_vsn
+  , Datamodel_common.quebec_release_schema_minor_vsn )
+
 let yangtze =
   ( Datamodel_common.yangtze_release_schema_major_vsn
   , Datamodel_common.yangtze_release_schema_minor_vsn
@@ -904,6 +908,26 @@ let upgrade_update_guidance =
       )
   }

+let fill_tunnel_protocol =
+  {
+    description=
+      "Fill up the new field protocol of a Tunnel"
+  ; version= (fun x -> x <= quebec)
+  ; fn=
+      (fun ~__context ->
+        Db.Tunnel.get_all ~__context
+        |> List.iter (fun self ->
+          let pif = Db.Tunnel.get_access_PIF ~__context ~self in
+          let network = Db.PIF.get_network ~__context ~self:pif in
+          let other_config = Db.Network.get_other_config ~__context ~self:network in
+          let encapsulation = List.assoc_opt "xo:sdn-controller:encapsulation" other_config
+            |> Option.value ~default:"gre" in
+          let value = Record_util.tunnel_protocol_of_string encapsulation in
+          Db.Tunnel.set_protocol ~__context ~self ~value
+        )
+      )
+  }
+
 let rules =
   [
     upgrade_domain_type
@@ -933,4 +957,5 @@ let rules =
   ; empty_pool_uefi_certificates
   ; upgrade_update_guidance
   ; upgrade_ca_fingerprints
+  ; fill_tunnel_protocol
   ]

 (* Maybe upgrade most recent db *)
